### DISCUSSION TITLE
initializing sqlitex.NewPool with per-connection pragmas is awkward

### DISCUSSION THREAD

**Discussion by rmmh** [zombiezen/go-sqlite#105] [Q&A] [answered]
A common optimization for WAL-mode databases is setting [PRAGMA synchronous=NORMAL](https://sqlite.org/pragma.html#pragma_synchronous), trading off some power-loss durability for ~50x more write throughput.

It's a per connection setting, so it has to be set for individual connection in the pool, with something like:

```go
	conns := []*sqlite.Conn{}
	for range poolsize {
		conn, _ := pool.Take(context.Background())
		conns = append(conns, conn)
		sqlitex.ExecuteTransient(conn, `PRAGMA synchronous=NORMAL`, nil)
	}

	for _, conn := range conns {
		pool.Put(conn)
	}
```

Or, shorter but depending on Pool being implemented with a buffered channel:

```go
	for range poolsize {
		conn, _ := pool.Take(context.Background())
		sqlitex.ExecuteTransient(conn, `PRAGMA synchronous=NORMAL`, nil)
		pool.Put(conn)
	}
```

There are other optimization pragmas that might be desired too, like `temp_store=MEMORY`.

Perhaps this could be wrapped in a helper function like `sqlitex.PoolExecAll(pool *Pool, query string) error` ?

**Comment by zombiezen** [comment, marked as answer]
This sort of every-connection setup is what [`PrepareConn`](https://pkg.go.dev/zombiezen.com/go/sqlite/sqlitex#PoolOptions.PrepareConn) is for. Does that work for your use case?

**Reply by rmmh** [reply]
Aha, I scanned the methods of pool and missed that part of PoolOptions, that works perfectly, thanks!
